### PR TITLE
Fix PRO metric calculation on GPU

### LINKED ISSUE #1316: [Bug]: PRO metric calculation fails on GPU
### Describe the bug

PRO metric fails on GPU if any of targets or predictions isn't in float type.

Error message:
```
RuntimeError: "max_pool2d_with_indices_out_cuda_frame" not implemented for 'Int'
```

### Dataset

MVTec

### Model

PADiM

### Steps to reproduce the behavior

1. Install anomalib from github
2. Run default padim training with PRO metric in config
3. When metrics are calculated the above mentioned error occurs

```
RuntimeError: "max_pool2d_with_indices_out_cuda_frame" not implemented for 'Int'
```

### OS information

OS information:
- OS: Windows 10
- Python version: 3.10
- Anomalib version: 0.7.0
- PyTorch version: 2.0.1
- CUDA/cuDNN version: 11.7
- GPU models and configuration: 1x gtx 1060
- Any other relevant information: /

### Expected behavior

Successful training and PRO metric calculation.

### Screenshots

_No response_

### Pip/GitHub

GitHub

### What version/branch did you use?

latest

### Configuration YAML

```yaml
dataset:
  name: mvtec
  format: mvtec
  path: ./datasets/MVTec
  category: bottle
  task: segmentation
  train_batch_size: 32
  eval_batch_size: 32
  num_workers: 0
  image_size: 256 # dimensions to which images are resized (mandatory)
  center_crop: null # dimensions to which images are center-cropped after resizing (optional)
  normalization: imagenet # data distribution to which the images will be normalized: [none, imagenet]
  transform_config:
    train: null
    eval: null
  test_split_mode: from_dir # options: [from_dir, synthetic]
  test_split_ratio: 0.2 # fraction of train images held out testing (usage depends on test_split_mode)
  val_split_mode: same_as_test # options: [same_as_test, from_test, synthetic]
  val_split_ratio: 0.5 # fraction of train/test images held out for validation (usage depends on val_split_mode)
  tiling:
    apply: false
    tile_size: null
    stride: null
    remove_border_count: 0
    use_random_tiling: False
    random_tile_count: 16

model:
  name: padim
  backbone: resnet18
  pre_trained: true
  layers:
    - layer1
    - layer2
    - layer3
  normalization_method: min_max # options: [none, min_max, cdf]

metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
    - PRO
  threshold:
    method: adaptive #options: [adaptive, manual]
    manual_image: null
    manual_pixel: null

visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: True # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]

project:
  seed: 42
  path: ./results
  unique_dir: True

logging:
  logger: [] # options: [comet, tensorboard, wandb, csv] or combinations.
  log_graph: false # Logs the model graph to respective logger.

optimization:
  export_mode: null # options: torch, onnx, openvino

# PL Trainer Args. Don't add extra parameter here.
trainer:
  enable_checkpointing: true
  default_root_dir: null
  gradient_clip_val: 0
  gradient_clip_algorithm: norm
  num_nodes: 1
  devices: 1
  enable_progress_bar: true
  overfit_batches: 0.0
  track_grad_norm: -1
  check_val_every_n_epoch: 1 # Don't validate before extracting features.
  fast_dev_run: false
  accumulate_grad_batches: 1
  max_epochs: 1
  min_epochs: null
  max_steps: -1
  min_steps: null
  max_time: null
  limit_train_batches: 1.0
  limit_val_batches: 1.0
  limit_test_batches: 1.0
  limit_predict_batches: 1.0
  val_check_interval: 1.0 # Don't validate before extracting features.
  log_every_n_steps: 50
  accelerator: auto # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">
  strategy: null
  sync_batchnorm: false
  precision: 32
  enable_model_summary: true
  num_sanity_val_steps: 0
  profiler: null
  benchmark: false
  deterministic: false
  reload_dataloaders_every_n_epochs: 0
  auto_lr_find: false
  replace_sampler_ddp: true
  detect_anomaly: false
  auto_scale_batch_size: false
  plugins: null
  move_metrics_to_cpu: false
  multiple_trainloader_mode: max_size_cycle
```

### Logs

```shell
File "anaconda3\envs\anomalib_env\lib\site-packages\pytorch_lightning\trainer\connectors\logger_connector\result.py", line 269, in compute
    return self.value.compute()
  File "anaconda3\envs\anomalib_env\lib\site-packages\torchmetrics\metric.py", line 534, in wrapped_func
    value = compute(*args, **kwargs)
  File "anomalib\src\anomalib\utils\metrics\pro.py", line 44, in compute
    comps = connected_components_gpu(target)
  File "anomalib\src\anomalib\utils\cv\connected_components.py", line 23, in connected_components_gpu
    components = connected_components(image, num_iterations=num_iterations)
  File "anaconda3\envs\anomalib_env\lib\site-packages\kornia\contrib\connected_components.py", line 54, in connected_components
    out[mask] = F.max_pool2d(out, kernel_size=3, stride=1, padding=1)[mask]
  File "anaconda3\envs\anomalib_env\lib\site-packages\torch\_jit_internal.py", line 484, in fn
    return if_false(*args, **kwargs)
  File "anaconda3\envs\anomalib_env\lib\site-packages\torch\nn\functional.py", line 782, in _max_pool2d
    return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
RuntimeError: "max_pool2d_with_indices_out_cuda_frame" not implemented for 'Int'
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by blaz-r**
## Description

Fixes PRO metric calculation on GPU by changing type to float, as well as matches shapes so it works with predictions of shape N, H, W.
I also adjusted tests to verify that it does works with relevant shapes and types.

- Fixes #1316 

## Changes

<details>
<summary>Describe the changes you made</summary>

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

</details>

## Checklist

<details>
<summary>Ensure that you followed the following</summary>

- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas (If applicable)
- [ ] I have made corresponding changes to the documentation (If applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (If applicable)

</details>

### CODE REVIEW

File: src/anomalib/utils/metrics/pro.py [hunk 1]
```diff
@@ -38,10 +38,12 @@ def compute(self) -> Tensor:
         target = dim_zero_cat(self.target)
         preds = dim_zero_cat(self.preds)
 
+        target = target.unsqueeze(1)  # kornia expects N1HW format
+        target = target.type(torch.float)  # kornia expects FloatTensor
```
Comment (samet-akcay) [inline src/anomalib/utils/metrics/pro.py:42]: Just to make it more Pythonic and idiomatic...
```suggestion
        target = target.unsqueeze(1).type(torch.float)  # kornia expects N1HW and FloatTensor format
```

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks for addressing this @blaz-r. I only have a single, very minor comment.

File: src/anomalib/utils/metrics/pro.py (hunk 1 above)
Comment (blaz-r) [reply src/anomalib/utils/metrics/pro.py:42]: Okay. I just wrote it as such since it's like that in aupro implementation. But I agree this is better :).

**Review by samet-akcay — APPROVED** [review]
lgtm, thanks!

**Pull request merged**